### PR TITLE
[Enhancement] Support column groups for final merge

### LINKED ISSUE #7332: Support column groups for final merge.
## Feature request

**Is your feature request related to a problem? Please describe.**

It will take up a lot of memory when merging temporary segment files in final merge procedure, which may cause OOM.

**Describe the solution you'd like**

Support column groups when merging segment files.

### PR DESCRIPTION
**Pull request by Linkerist**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7332 

## Problem Summary(Required) ：

It will take up a lot of memory when merging temporary segment files in final merge procedure, which may cause OOM.
Support column groups when merging segment files.
There does exists some repetitive code, but other unrelated code will be influenced if refactor it now, so let them go and refactor it in next individual PR.

reduce memory 13~16 times

a case like this one(5000 columns, 4.3G, 100000 row), other data scales are almost same ratio

before this PR:
heap1

after this PR:
heap

36% performance improvement

before this PR:
19376ms
profile1

after this PR:
14203ms
profile

### CODE REVIEW

**Comment by trueeyu** [comment]
Better to test the memory optimization effect and performance optimization effect

**Comment by Linkerist** [comment]
run starrocks_be_unittest

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 1]
```diff
@@ -425,107 +431,328 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
...
+            auto segment_ptr = Segment::open(ExecEnv::GetInstance()->tablet_meta_mem_tracker(), _fs, tmp_segment_file,
+                                             seg_id, _context.tablet_schema);
+            if (!segment_ptr.ok()) {
+                LOG(WARNING) << "Fail to open " << tmp_segment_file << ": " << segment_ptr.status();
+                return segment_ptr.status();
+            }
+            if ((*segment_ptr)->num_rows() == 0) {
+                continue;
+            }
+            auto res = (*segment_ptr)->new_iterator(schema, seg_options);
```
Comment (chaoyli) [inline be/src/storage/rowset/beta_rowset_writer.cpp:465]: Why creating an iterator will encounter an end_of_file?

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 2]
```diff
@@ -425,107 +431,328 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
...
+                  << " #key:" << _context.tablet_schema->num_key_columns() << " input("
+                  << "entry=" << seg_iterators.size() << " rows=" << stats.raw_rows_read
+                  << " bytes=" << PrettyPrinter::print(stats.bytes_read, TUnit::UNIT) << ") output(rows=" << total_rows
+                  << " chunk=" << total_chunk << " bytes=" << PrettyPrinter::print(total_data_size(), TUnit::UNIT)
+                  << ") duration: " << timer.elapsed_time() / 1000000 << "ms";
     } else {
-        itr = new_aggregate_iterator(new_heap_merge_iterator(seg_iterators), 0);
-    }
-    itr->init_encoded_schema(vectorized::EMPTY_GLOBAL_DICTMAPS);
+        auto schema = vectorized::ChunkHelper::convert_schema_to_format_v2(*_context.tablet_schema);
```
Comment (chaoyli) [inline be/src/storage/rowset/beta_rowset_writer.cpp:650]: I have not found many differences between the if/else and so many redundant codes.

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 3]
```diff
@@ -425,107 +431,328 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
...
+        for (int seg_id = 0; seg_id < _num_segment; ++seg_id) {
+            std::string tmp_segment_file =
+                    BetaRowset::segment_temp_file_path(_context.rowset_path_prefix, _context.rowset_id, seg_id);
+            auto segment_ptr = Segment::open(ExecEnv::GetInstance()->tablet_meta_mem_tracker(), _fs, tmp_segment_file,
+                                             seg_id, _context.tablet_schema);
+            if (!segment_ptr.ok()) {
+                LOG(WARNING) << "Fail to open " << tmp_segment_file << ": " << segment_ptr.status();
+                return segment_ptr.status();
+            }
+            if ((*segment_ptr)->num_rows() == 0) {
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/rowset/beta_rowset_writer.cpp]: changed in https://github.com/StarRocks/starrocks/pull/7333/commits/9b599bd408c4379fe318a8f0672d4eb67e1ba3d8

File: be/src/storage/rowset/beta_rowset_writer.cpp (hunk 1 above)
Comment (Linkerist) [reply be/src/storage/rowset/beta_rowset_writer.cpp:465]: changed in https://github.com/StarRocks/starrocks/pull/7333/commits/7c6892471b5664332640c65e28debedf40c09b91

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -425,107 +431,323 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
     MonotonicStopWatch timer;
     timer.start();
 
-    auto schema = vectorized::ChunkHelper::convert_schema_to_format_v2(*_context.tablet_schema);
+    if (CompactionUtils::choose_compaction_algorithm(_context.tablet_schema->num_columns(),
+                                                     config::vertical_compaction_max_columns_per_group,
+                                                     _num_segment) == VERTICAL_COMPACTION) {
```
Comment (wyb) [inline be/src/storage/rowset/beta_rowset_writer.cpp:436]: `seg_iterators` does not contain the segment that has no rows.
`_num_segment` (`source_num`) should to be same with `seg_iterators`.

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -425,107 +431,323 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
...
+        auto source_masks = std::make_unique<std::vector<vectorized::RowSourceMask>>();
+
+        ChunkIteratorPtr itr;
+        // create temporary segment files at first, then merge them and create final segment files if schema change with sorting
+        if (_context.write_tmp) {
+            if (_context.tablet_schema->keys_type() == KeysType::DUP_KEYS) {
+                itr = new_mask_merge_iterator(seg_iterators, mask_buffer.get());
+            } else if (_context.tablet_schema->keys_type() == KeysType::UNIQUE_KEYS ||
+                       _context.tablet_schema->keys_type() == KeysType::AGG_KEYS) {
+                itr = new_aggregate_iterator(new_mask_merge_iterator(seg_iterators, mask_buffer.get()), true);
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/rowset/beta_rowset_writer.cpp]: sorry, i forgot to change this branch as well. changed.

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 6]
```diff
@@ -425,107 +431,323 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
...
-            itr = new_aggregate_iterator(new_heap_merge_iterator(seg_iterators), 0);
-        } else {
-            return Status::NotSupported(fmt::format("HorizontalBetaRowsetWriter not support {} key type final merge",
-                                                    _context.tablet_schema->keys_type()));
+        // since the segment already NONOVERLAPPING here, make the _create_segment_writer
+        // method to create segment data files, rather than temporary segment files.
+        _context.segments_overlap = NONOVERLAPPING;
+
+        _vertical_beta_rowset_writer = std::make_unique<VerticalBetaRowsetWriter>(_context);
+        _vertical_beta_rowset_writer->init();
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/rowset/beta_rowset_writer.cpp]: changed

File: be/src/storage/rowset/beta_rowset_writer.cpp (hunk 4 above)
Comment (Linkerist) [reply be/src/storage/rowset/beta_rowset_writer.cpp]: changed.

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 7]
```diff
@@ -437,95 +440,305 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
...
+        _context.max_rows_per_segment = CompactionUtils::get_segment_max_rows(config::max_segment_file_size,
+                                                                              _num_rows_written, _total_data_size);
+
+        auto chunk_shared_ptr = vectorized::ChunkHelper::new_chunk(schema, config::vector_chunk_size);
+        auto chunk = chunk_shared_ptr.get();
+
+        _num_segment = 0;
+        _num_delfile = 0;
+        _num_rows_written = 0;
+        _num_rows_del = 0;
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/rowset/beta_rowset_writer.cpp:511]: we do not support mixed upsert/delete for multi-segment, these member variable will always be zero if code can run here, set it to zero, if we support mixed upsert/delete, they can be update in vertical writer

**Review by wyb — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**